File: WEEK7/Js/bj6_1915.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const [n, m] = input[0].split(' ').map(Number);
const dp = Array.from({ length: n + 1 }, () => new Int32Array(m + 1));
let maxLen = 0;

for (let i = 1; i <= n; i++) {
    const line = input[i];
    for (let j = 1; j <= m; j++) {
        if (line[j - 1] === '1') {
            dp[i][j] = Math.min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]) + 1;
            maxLen = Math.max(maxLen, dp[i][j]);
        }
    }
}

console.log(maxLen * maxLen);

File: WEEK7/Python/bj6_1915.py
import sys

n, m = map(int, sys.stdin.readline().split())
dp = [[0] * (m + 1) for _ in range(n + 1)]
max_len = 0

for i in range(1, n + 1):
    line = sys.stdin.readline().strip()
    for j in range(1, m + 1):
        if line[j - 1] == "1":
            dp[i][j] = min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]) + 1
            max_len = max(max_len, dp[i][j])

print(max_len * max_len)
